sqlite/ext/comdb2/tags.c

```diff
@@ -0,0 +1,156 @@
+    char *tablename;
+    char *tagname;
+    int ixnum;
+    int64_t recsize;
+    int64_t nmembers;
+    char *sqlitetag;
+    char *csctag;
+    int numblobs;
+    int64_t nix;
+};
+
+int get_tags_count(int columns)
+{
+    sqlite3_int64 tableid = 0;
+    int ntags;
+
+    for (tableid = 0, ntags = 0;
+            comdb2_next_allowed_table(&tableid) == SQLITE_OK && tableid < thedb->num_dbs;
+            ++tableid) {
+        ntags += get_table_tags_count(thedb->dbs[tableid]->tablename, columns);
+    }
+
+    return ntags;
+}
+
+int get_tags_details(struct systable_tag *data, int ntags)
+{
+    sqlite3_int64 tableid;
+    int sum, rc;
+    struct dbtable *table;
+
+    for (tableid = 0, sum = 0;
+            comdb2_next_allowed_table(&tableid) == SQLITE_OK &&
+            tableid < thedb->num_dbs && sum < ntags;
+            ++tableid) {
+        table = thedb->dbs[tableid];
+
+        rc = get_table_tags(table->tablename, &data[sum], 0, ntags - sum);
+        if (rc < 0)
+            return -1;
+        sum += rc;
+    }
+
+    return 0;
+}
+
+static int collect_tags(void **pd, int *pn)
+{
+    struct systable_tag *data = NULL;
+    int ntags, rc;
+
+    /* Get number of rows, and allocate a continuous chunk of memory. */
+    ntags = get_tags_count(0);
+    data = calloc(ntags, sizeof(struct systable_tag));
+     
+    rc = get_tags_details(data, ntags);
+    if (rc < 0) {
+        *pd = NULL;
+        *pn = 0;
+    } else {
+        *pd = data;
+        *pn = ntags;
+    }
+    return rc;
+}
+
+
+int systable_set_tag(struct systable_tag *parr, int num, const char * tblname,
+                     const struct  schema *s)
+{
+    struct systable_tag *p = &parr[num];
+    p->tablename = strdup(tblname);
+    p->tagname = strdup(s->tag);
+    p->ixnum = s->ixnum;
+    p->recsize = s->recsize;
+    p->nmembers = s->nmembers;
+    p->sqlitetag = strdup(s->sqlitetag ? s->sqlitetag : "");
+    p->csctag = strdup(s->csctag ? s->csctag : "");
+    p->numblobs = s->numblobs;
+    p->nix = s->nix;
+    return 1;
+}
+
+static void free_tags(void *data, int n)
+{
+    struct systable_tag *tags= data, *p;
+    int i;
+    for (i = 0; i < n; ++i) {
+        p = &tags[i];
+        free(p->tablename);
+        free(p->tagname);
+        free(p->sqlitetag);
+        free(p->csctag);
+    }
+    free(tags);
+}
+
+sqlite3_module systblTagsModule = {
+  .access_flag = CDB2_ALLOW_ALL,
+  .systable_lock = "comdb2_tables"
+};
+
+int systblTagsInit(sqlite3 *db)
+{
+    return create_system_table(
+        db, "comdb2_tags", &systblTagsModule,
+        collect_tags, free_tags, sizeof(struct systable_tag),
+        CDB2_CSTRING, "tablename", -1, 0,
+        CDB2_CSTRING, "tagname", -1, offsetof(struct systable_tag, tagname),
+        CDB2_INTEGER, "ixnum", -1, offsetof(struct systable_tag, ixnum),
+        CDB2_INTEGER, "size", -1, offsetof(struct systable_tag, recsize),
+        CDB2_INTEGER, "columns", -1, offsetof(struct systable_tag, nmembers),
```

Comment:
perhaps numcolumns ? to keep consistent with numblobs and numindexes below ?

Reply:
done